core/base-service/base-graphql.js

```diff
@@ -0,0 +1,48 @@
+'use strict'
+
+const { parse } = require('graphql')
+const BaseJsonService = require('./base-json')
+const { InvalidResponse } = require('./errors')
+
+function defaultGraphqlErrorHandler(errors) {
+  throw new InvalidResponse({ prettyMessage: errors[0].message })
+}
+
+class BaseGraphqlService extends BaseJsonService {
+  _validateQuery(query) {
+    // Attempting to parse the query string
+    // will throw a descriptive exception if it isn't valid
+    parse(query)
+  }
+
+  async _requestGraphql({
+    schema,
+    url,
+    query,
+    variables = {},
+    options = {},
+    httpErrorMessages = {},
+    graphqlErrorHandler = defaultGraphqlErrorHandler,
```

Comment:
In general, Error handling for graphql is going to require a mixed approach:

* `httpErrorMessages`: Certain errors (e.g: `401 Unauthorized`, `500 Internal Server Error`) are going to be returned as status codes so we do need to be able to pass these.
* `graphqlErrorMessageHandler`: With a graphql, a single request can do several things. Some might error, some might not so we need to handle errors returned in an `errors` object (possibly accompanied by a partial valid response) with a `200 OK` status. All the spec says is that an error needs a `.message` property, so these are going to be a bit of a mixed bag from API to API.

Reply:
It seems like the job of this function is to take an `errors` object and (hopefully) throw an error. I wonder if we could set this up instead as a `transformErrors` function, which gets an errors object and is responsible for returning an instance of `ShieldsRuntimeError`. We could verify that an error is returned, and then throw it. It reduces the chance that a bad error handler accidentally lets control slip back to the handler. Since there is not really much to do besides translate the errors object into a Shields error, I like the idea of calling it transform, which is more narrowly scoped than handle.

Reply:
I see where you're going with that. I've phrased my follow-up question as a code snippet:

```js
if (json.errors) {
  const exception = transformErrors(json.errors)
  if (exception instanceof ShieldsRuntimeError) {
    throw exception
  } else {
    // what happens here?
    // do we throw a runtime error (i.e: log it to sentry if it happens in prod)
    // or throw a generic ShieldsRuntimeError (i.e: render an error badge if it happens in prod)
  }
}
```

Reply:
I think it can should work like an assertion:

```js
throw Error(`transformErrors() must return a ShieldsRuntimeError; got ${exception}`)
```

In development or service tests that will bubble right up, and in production it will get handled as an internal error.